server/bin/state/test-30.1.config/pbench-server.cfg

```diff
@@ -0,0 +1,32 @@
+install-dir = %(unittest-dir)s/opt/pbench-server
+
+###########################################################################
+## Deployment section
+###########################################################################
+[pbench-server]
+pbench-top-dir = %(unittest-dir)s/pbench
+pbench-local-dir = %(unittest-dir)s/pbench-local
+pbench-backup-dir = %(pbench-local-dir)s/archive.backup
+# Add role for sync'ing with satellites
+debug_unittest = True
+commit_id = unit-test
+environment = unit-test
+admin-email = unit-test-user@example.com
+put-token = Authentication-token
+
+[results]
+server_rest_url = http://pbench.example.com/api/v1/
```

Comment:
Again; this is useful for testing, but shouldn't be submitted.

On the other hand, the "gold" unit now need this. Two options occur to me:

1. The unittests script already creates a `pbench-server.cfg` file by copying the template and adding data; you could add a `[results]` section then, strictly for the sake of the unit testing; (look in `_setup_state`), or,
2. You could mock the code that gets the option value rather than requiring it to be in the template file.

Reply:
Also this section is a part of agent config. On the server side the equivalent is a `rest_uri` option under `pbench-server` section, can that be used here?

Reply:
My understanding is that the put-shim will operate _as a Pbench Server_, but only during the 0.69-0.72 transition.  It will accept uploads from 0.69 Agent clients, and then forward them to a 0.71/0.72 Server.  So, it will read a "server configuration file", because it's a server, but it will need to get from that where an 0.72 Agent _would_ have sent an upload.  So, rather than having it _also_ read an Agent configuration file, we just need to make sure that it can somehow find what it needs in the Server configuration file.  And, as Nikhil pointed out, that information is in the Server file, just in a different location from where it is in the Agent file.

The reason that that information is in the Server file is (I believe...) that the Server needs to know what URL it is serving.  (But, now that you've asked the question, I'm not sure _why_ it needs to know that, since the requests will still come to it, even if _it_ doesn't know where it is....)  Perhaps it is to resolve the question of which address to bind the listener to when deployed on a machine with multiple NICs...I don't know.

As for packaging and deployment, my expectation is that the put-shim is basically one file which we will include with the Server package, during the 0.69-0.72 transition (and then we'll remove it from the repo).  As I said, it depends on other modules in the the Server code, so, in order to deploy the put-shim, we need (most of) the rest of the Server.  (However, the normal Server deployment doesn't need the put-shim code....)  I expect there will only ever be one put-shim deployment, so we can hand-craft the configuration file for it when it is installed, and I don't think it will be worthwhile to tease out the individual dependencies from the rest of the Server code -- we should just install the Server RPM with the put-shim riding along, modify the config file appropriately, and invoke the put-shim instead of the normal Server.

Reply:
@ndokos `rest_uri` is for registering API endpoints its defined like this in the config file `/api/v%(rest_version)s`